---
layout: post
title: 예외처리
categories: [java]
tags: [java]
created-time: "20:42"
---

### Java 에서는 예외를 두 가지로 분류한다.

**1. 체크 예외 (Checked Exception)**

- `try-catch-finally` 로 예외를 처리하지 않으면 컴파일 오류가 발생한다.
- `RuntimeException` 을 제외한 `Exception` 클래스의 자식 클래스

**2. 언체크 예외 (Unchecked Exception)**

- 런타임 시에 예외가 발생하기 때문에 컴파일 시 오류가 발생하지 않고 예외가 발생하지 않도록 주의해야 한다.
- `RuntimeException` 클래스의 자식 클래스

![예외처리 클래스의 상속관계](/assets/img/posts/java/java3_1.png)

### try-catch-finally 와 throws 의 차이

`try-catch-finally` 구문은 예외를 처리하고 프로그램의 흐름을 제어할 때 사용하고, `throws` 구문은 메소드에서 발생할 수 있는 예외를 호출한 쪽으로 알릴 때 사용한다.

둘의 차이점은 예외를 직접 처리하는지 아니면 예외처리를 호출한 쪽으로 위임하는지 이다.

특히 `throws` 는 트랜잭션과 밀접한 관련이 있다.

{: .box-note}
💡 트랜잭션은 하나의 작업 단위를 뜻한다.

예를 들어 한 트랜잭션이 A, B, C 라는 작업으로 이루어져 있을 때 A, B, C 중 예외가 발생하면 트랜잭션을 실행하기 전 단계로 되돌려야 할 것이다.

이 때 `throws` 를 이용해서 예외처리를 할 수 있다.

```java
void transaction() {
    try {
        A();
        B();
        C();
    } catch(Exception e) {
        // reset
    }
}

void A() throws Exception {
}

void B() throws Exception {
}

void C() throws Exception {
}
```

### try-with-resources 문

Java 7 부터는 사용한 자원을 자동으로 해제해주는 `try-with-resources` 문을 사용할 수 있다.

```java
...
try(BufferedReader br = new BufferedReader(new FileReader(filePath))) {
    return br.readLine();
}
...
```